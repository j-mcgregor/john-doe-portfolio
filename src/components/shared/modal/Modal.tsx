import React, { useEffect } from 'react'

export interface ModalProps {
    src: string
    isOpen: boolean
    handleClose(): void
}

export interface ContentProps {
    children?: React.ReactNode
    onClick(): void
}

export const Content: React.FC<ContentProps> = ({ children, onClick }) => {
    // Refs are used when we want to interact / modify a React element that may be outside of our render
    // Here, we're mapping all of the children passed into this component so we can easily find it on the DOM
    const refs = React.Children.map(children, () =>
        React.createRef<HTMLDivElement>()
    )

    // Whenever the document receives a click element, we want to trigger this method
    const handleClick = (e: MouseEvent) => {
        // no children, no refs
        if (refs) {
            // for each ref (child)
            const isOutside = refs.every(ref => {
                // https://stackoverflow.com/questions/43842057/detect-if-click-was-inside-react-component-or-not-in-typescript
                return (
                    ref &&
                    ref.current &&
                    !ref.current.contains(e.target as Node)
                )
            })

            if (isOutside) {
                onClick()
            }
        }
    }

    useEffect(() => {
        document.addEventListener('click', handleClick)

        return () => {
            document.removeEventListener('click', handleClick)
        }
    })

    const mapped = React.Children.map(children, (child, x) => {
        // https://stackoverflow.com/questions/42261783/how-to-assign-the-correct-typing-to-react-cloneelement-when-giving-properties-to
        return (
            refs &&
            React.isValidElement(child) &&
            React.cloneElement(child, {
                ref: refs[x],
            })
        )
    })

    return <React.Fragment>{mapped}</React.Fragment>
}

const Modal: React.FC<ModalProps> = ({ isOpen, src, handleClose }) => {
    return (
        <div className="modal" style={{ display: isOpen ? 'block' : 'none' }}>
            <span className="close" onClick={handleClose} data-testid="close">
                &times;
            </span>
            <Content onClick={handleClose}>
                <img className="modal-content" src={src} />
            </Content>
            <div className="caption"></div>
        </div>
    )
}

export default Modal
