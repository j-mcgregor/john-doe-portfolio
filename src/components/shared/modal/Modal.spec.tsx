import * as React from 'react'
import { render, act } from '@testing-library/react'
import Modal, { ModalProps, ContentProps, Content } from './Modal'

describe('Modal', () => {
    let props: ModalProps
    beforeEach(() => {
        props = {
            isOpen: true,
            src: 'http://fillmurrary.com/500/500',
            handleClose: jest.fn(),
        }
    })

    it('should show display: "block" when isOpen is true', () => {
        const { baseElement } = render(<Modal {...props} />)

        expect(baseElement).toMatchSnapshot()
    })
    it('should show display: "none" when isOpen is false', () => {
        const { baseElement } = render(<Modal {...props} isOpen={false} />)

        expect(baseElement).toMatchSnapshot()
    })

    it('should close the modal when the X is clicked', () => {
        const { getByTestId } = render(<Modal {...props} />)

        const res = getByTestId('close')

        act(() => {
            res.click()
        })
        expect(props.handleClose).toHaveBeenCalled()
    })
})

describe('Content', () => {
    let props: ContentProps

    beforeEach(() => {
        props = {
            onClick: jest.fn(),
        }
    })

    // want to test that onClick is firing when clicking outside
    it('should not trigger onClick when child is clicked', () => {
        const { getByTestId } = render(
            <Content {...props}>
                <div data-testid="inside">Test</div>
            </Content>
        )

        const res = getByTestId('inside')

        act(() => {
            res.click()
        })

        expect(props.onClick).not.toHaveBeenCalled()
    })

    it('should trigger onClick when outside child is clicked', () => {
        const { getByTestId } = render(
            <div data-testid="outside">
                <Content {...props}>
                    <div>Test</div>
                </Content>
            </div>
        )

        const res = getByTestId('outside')

        act(() => {
            res.click()
        })

        expect(props.onClick).toHaveBeenCalled()
    })

    it('should do nothing if no children are passed', () => {
        const { getByTestId } = render(
            <div data-testid="outside">
                <Content {...props}></Content>
            </div>
        )

        const res = getByTestId('outside')

        act(() => {
            res.click()
        })

        expect(props.onClick).not.toHaveBeenCalled()
    })

    it('should do nothing if no children are passed 2', () => {
        jest.spyOn(React.Children, 'map').mockReturnValue([])

        const { queryByTestId } = render(
            <div data-testid="outside">
                <Content {...props}>
                    <div data-testid="inside-XXX"></div>
                </Content>
            </div>
        )

        const res = queryByTestId('inside-XXX')

        act(() => {
            res && res.click()
        })

        expect(props.onClick).not.toHaveBeenCalled()
    })
})
