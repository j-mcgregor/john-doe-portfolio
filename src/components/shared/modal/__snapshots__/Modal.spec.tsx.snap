// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should show display: "block" when isOpen is true 1`] = `
<body>
  <div>
    <div
      class="modal"
      style="display: block;"
    >
      <span
        class="close"
        data-testid="close"
      >
        ×
      </span>
      <img
        class="modal-content"
        src="http://fillmurrary.com/500/500"
      />
      <div
        class="caption"
      />
    </div>
  </div>
</body>
`;

exports[`Modal should show display: "none" when isOpen is false 1`] = `
<body>
  <div>
    <div
      class="modal"
      style="display: none;"
    >
      <span
        class="close"
        data-testid="close"
      >
        ×
      </span>
      <img
        class="modal-content"
        src="http://fillmurrary.com/500/500"
      />
      <div
        class="caption"
      />
    </div>
  </div>
</body>
`;
